interfaces/cython/cantera/test/test_jacobian.py

```diff
@@ -0,0 +1,708 @@
+        #   species: [H2, H, O, O2, OH, H2O, HO2, H2O2, AR, N2]
+        cls.gas.X = [0.1, 1e-4, 1e-5, 0.2, 2e-4, 0.3, 1e-6, 5e-5, 0.3, 0.1]
+        cls.gas.TP = 800, 2 * ct.one_atm
+        super().setUpClass()
+
+
+class TestElementaryRev(HydrogenOxygen, utilities.CanteraTest):
+    # Standard elementary reaction with two reactants
+    # - equation: O + H2 <=> H + OH  # Reaction 3
+    #   rate-constant: {A: 3.87e+04, b: 2.7, Ea: 6260.0}
+    rxn_idx = 2
+
+
+class TestElementarySelf(HydrogenOxygen, utilities.CanteraTest):
+    # Elementary reaction with reactant reacting with itself
+    # - equation: 2 HO2 <=> O2 + H2O2  # Reaction 28
+    rxn_idx = 27
+
+
+class TestFalloff(HydrogenOxygen, utilities.CanteraTest):
+    # Fall-off reaction
+    # - equation: 2 OH (+M) <=> H2O2 (+M)  # Reaction 22
+    #   type: falloff
+    #   low-P-rate-constant: {A: 2.3e+18, b: -0.9, Ea: -1700.0}
+    #   high-P-rate-constant: {A: 7.4e+13, b: -0.37, Ea: 0.0}
+    #   Troe: {A: 0.7346, T3: 94.0, T1: 1756.0, T2: 5182.0}
+    #   efficiencies: {H2: 2.0, H2O: 6.0, AR: 0.7}
+    rxn_idx = 21
+    rtol = 1e-4
+
+
+class TestThreeBody(HydrogenOxygen, utilities.CanteraTest):
+    # Three body reaction with default efficiency
+    # - equation: O + H + M <=> OH + M  # Reaction 2
+    #   type: three-body
+    #   rate-constant: {A: 5.0e+17, b: -1.0, Ea: 0.0}
+    #   efficiencies: {H2: 2.0, H2O: 6.0, AR: 0.7}
+    rxn_idx = 1
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.ix3b = list(range(cls.gas.n_species))
+
+    def test_thirdbodies_forward(self):
+        drop = self.gas.forward_rop_species_derivatives
+        self.gas.jacobian_settings = {"skip-third-bodies": True}
+        drops = self.gas.forward_rop_species_derivatives
+        dropm = drop - drops
+        rop = self.gas.forward_rates_of_progress
+        self.assertNear(rop[self.rxn_idx], (dropm[self.rxn_idx] * self.gas.X).sum())
+
+    def test_thirdbodies_reverse(self):
+        drop = self.gas.reverse_rop_species_derivatives
+        self.gas.jacobian_settings = {"skip-third-bodies": True}
+        drops = self.gas.reverse_rop_species_derivatives
+        dropm = drop - drops
+        rop = self.gas.reverse_rates_of_progress
+        self.assertNear(rop[self.rxn_idx], (dropm[self.rxn_idx] * self.gas.X).sum())
+
+
+class EdgeCases(RateExpressionTests):
+
+    @classmethod
+    def setUpClass(cls):
+        cls.gas = ct.Solution("jacobian-tests.yaml", transport_model=None)
+        #   species: [H2, H, O, O2, OH, H2O, HO2, H2O2, AR]
+        cls.gas.X = [0.1, 1e-4, 1e-5, 0.2, 2e-4, 0.3, 1e-6, 5e-5, 0.4]
+        cls.gas.TP = 800, 2 * ct.one_atm
+        super().setUpClass()
+
+
+class TestElementaryIrr(EdgeCases, utilities.CanteraTest):
+    # Irreversible elementary reaction with two reactants
+    # - equation: O + HO2 => OH + O2  # Reaction 1
+    #   rate-constant: {A: 2.0e+13, b: 0.0, Ea: 9.252008e+04}
+    rxn_idx = 0
```

Comment:
Rather than having to duplicate the reaction definitions here as a means of documentatation, would it make more sense to just define the parameters of the `Reaction` and use those to instantiate a phase with that one reaction?

Reply:
The way these tests are implemented, all reactions except one are 'switched off', while keeping the original size of the coefficient matrix. That allows for checking that all indices are assigned correctly, which is imho a stricter / more reliable test than assigning a single reaction for a given phase.

While I see your point, the docstring is purely a reminder of what is done, and could in theory be shortened: all the information is easily accessible regardless.